ts/private/ts_project_worker.js

```diff
@@ -0,0 +1,384 @@
+            for (const [path, content] of emittedFiles.entries()) {
+                const sourcePath = emittingMap.get(path);
+                if(!builder.getSourceFile(sourcePath)) {
+                    debuglog(`removing ${path} from emit cache`);
+                    emittedFiles.delete(path);
+                    emittingMap.delete(path);
+                } else {
+                    writeFile(path, content);
+                }
+            }
+        }
+
+        const writeF = (fileName, data, writeByteOrderMark, onError, sourceFiles) => {
+            writeFile(fileName, data, writeByteOrderMark, onError, sourceFiles);
+            emittingMap.set(fileName, sourceFiles[0].fileName)
+            emittedFiles.set(fileName, data)
+        }
+        return emit(targetSourceFile, writeF, cancellationToken, emitOnlyDtsFiles, customTransformers)
+    }
+
+    return builder;
+}
+
+/** @type {ts.FormatDiagnosticsHost} */
+const formatDiagnosticHost = {
+    getCanonicalFileName: (path) => path,
+    getCurrentDirectory: ts.sys.getCurrentDirectory,
+    getNewLine: () => ts.sys.newLine,
+}
+
+function printDiagnostics(diagnostics) {
+    worker.log('')
+    worker.log(ts.formatDiagnostics(diagnostics, formatDiagnosticHost))
+}
+
+
+
+function isAncestorDirectory(ancestor, child) {
+    const ancestorChunks = ancestor.split(path.sep).filter(i => !!i);
+    const childChunks = child.split(path.sep).filter(i => !!i);
+    return ancestorChunks.every((chunk, i) => childChunks[i] === chunk)
+}
+
+/** @type {Map<string, {program: ts.EmitAndSemanticDiagnosticsBuilderProgram, host: ts.WatchCompilerHost, previousInputs?: import("@bazel/worker").Inputs}>} */
+const workers = new Map(); 
+
+function getTsConfigPath(args) {
+    return args[args.indexOf('--project') + 1]
+}
+
+function createProgram(args, initialInputs) {
+    const tsconfig = getTsConfigPath(args);
+    const execRoot = path.resolve(process.cwd(), '..', '..', '..')
+
+    debuglog(`tsconfig: ${tsconfig}`);
+    debuglog(`execroot ${execRoot}`);
+
+    const parsedArgs = ts.parseCommandLine(args)
+
+    const directoryWatchers = new Map();
+    const fileWatchers = new Map();
+
+    const knownInputs = new Set(initialInputs);
+
+    /** @type {ts.System} */
+    const sys = {
+        ...ts.sys,
+        write: worker.log,
+        writeOutputIsTTY: false,
+        setTimeout: setImmediate,
+        clearTimeout: clearImmediate,
+        fileExists: fileExists,
+        readFile: readFile,
+        writeFile: writeFile,
+        readDirectory: readDirectory,
+        directoryExists: directoryExists,
+        getDirectories: getDirectories,
+        watchFile: watchFile,
+        watchDirectory: watchDirectory
+    }
+
+    const host = ts.createWatchCompilerHost(
+        tsconfig,
+        parsedArgs.options,
+        sys,
+        createEmitCacheAndDiagnosticsProgram,
+        noop,
+        noop
+    )
+
+    host.invalidate = invalidate;
+
+    program = ts.createWatchProgram(host)
```

Comment:
Where is `program` defined?

Reply:
Looks like it's not defined anywhere (it's a lint error if you run a linter), so this is creating + modifying a global. If this worker is ever used concurrently and there's async callbacks this would very likely cause bugs.

Reply:
Ah.. good catch!